rfcs/shared/build-system/08-type-checking-perf-improvements.md

```diff
@@ -0,0 +1,287 @@
+
+1. generate problematic(circular) dependencies types(`.d.ts`) first manually before `type-check` runs
+2. get rid of circular dependencies
+
+### 1. generate problematic(circular) dependencies types(`.d.ts`) first manually before `type-check` runs
+
+We can imperatively traverse graph and generate `d.ts` for packages that cause circular dependencies as can bee seen in this [PR](https://github.com/microsoft/fluentui/pull/28002/files#diff-38d7f0411164173eb390d7191ad068bfe70935082bc37198b6b35b00408b2b42).
+
+This dts generation can be done in 2 ways:
+
+**1. generate tasks first, then run builds**
+
+> EXAMPLE Logic: https://github.com/microsoft/fluentui/pull/28002/files#diff-7915b9b726a397ae7ba6af7b9703633d21c031ebf21682f3ee7e6a4ec52837a5
+>
+> EXAMPLE PR: https://github.com/microsoft/fluentui/pull/28002
+
+```yml
+- script: node ./scripts/executors/type-check-ci-hack.js --base $(targetBranch)
+    displayName: type-check perf preparation
+- script: yarn buildci $(sinceArg)
+```
+
+**Pros:**
+
+- none
+
+**Cons:**
+
+- bad DX for local machine
+
+**2. Leverage task graph orchestration and define dependency on every v9 `type-check`**
+
+> EXAMPLE Logic: https://github.com/microsoft/fluentui/pull/30577/files#diff-624201bdc18cf7b31fdca43ca08549cbd4db08cf5fa3fc9dfb51b0141669b1b8
+>
+> EXAMPLE PR: https://github.com/microsoft/fluentui/pull/30577
+
+```jsonc
+// @filename <package-root>/project.json
+{
+  "targets": {
+    "type-check": {
+      "dependsOn": [
+        /* global nx task/script that will be run automatically - its cached so it will do work only once  */
+        "generate-circular-deps-dts"
+      ]
+    }
+  }
+}
+```
+
+With this the flow would be
+
+```mermaid
+flowchart LR
+ type-check --> generate-circular-deps-dts --> build[^build]
+```
+
+**Pros:**
+
+- declarative approach
+- same behaviour on CI and local machine (NOTE: we run into race-conditions within MVP)
+
+**Cons:**
+
+- all problematic packages needs to override `targets#type-check` with `dependsOn` declaration
+- problems with current task orchestrator caching
+- huge peak of CPU usage
+
+#### Solution overall Pros and Cons
+
+**Pros:**
+
+- no package scaffold changes needed
+
+**Cons:**
+
+- brittle solution for future library requirements
+- manual processing and resolution of dependency graph + maintenance
+- adding more technical depth and magic on build layer
+- "hack" that is not solving circular dependencies
+
+### 2. get rid of circular dependencies
+
+We have 2 options on how to get rid of circular dependencies
+
+- stop importing from react-components suite within stories
+- move/encapsulate problematic parts of code to separate packages.
+
+#### 2.1 stop importing from react-components suite within stories
+
+```diff
+-import {Button} from '@fluentui/react-components'
++import {Button} from '@fluentui/react-button'
+```
+
+##### Solution overall Pros and Cons
+
+**Pros:**
+
+- no magic resolution of dependency tree behind the scenes
```

Comment:
~~This option should also result in a smaller package graph?~~

No, it will not as we have suite package `@fluentui/react-components` and we will have to build all dependencies anyway, for example:

```json5
// packages/react-image.json
{
  "name": "@fluentui/react-image",
  "dependencies": {}
}
```

```json5
// packages/react-button.json
{
  "name": "@fluentui/react-button",
  "dependencies": {}
}
```

```json5
// packages/react-popover.json
{
  "name": "@fluentui/react-popover",
  "dependencies": {},
  "devDependencies": {
    "@fluentui/react-button": "*"
  }
}
```

```json5
// packages/react-components.json
{
  "name": "@fluentui/react-components",
  "dependencies": {
    "@fluentui/react-button": "*",
    "@fluentui/react-image": "*",
    "@fluentui/react-popover": "*"
  }
}
```

With that layout anyway all component packages will be built as `@fluentui/react-components` requires them.

----

**However**, _hypothetically_  this option should be better for task orchestration as for 2.2.1 & 2.2.2 packages like `@fluentui/react-image-storybook` won't be able to run scripts before `@fluentui/react-components` will be built. That creates a bottleneck in build process:

```mermaid
flowchart LR
    FU[fluentui/react-utilities] --> FI
    FU --> FB
    FU --> FP
    FU --> FC
    FI[fluentui/react-image] --> FC[fluentu/react-components]
    FB[fluentui/react-button] --> FC
    FP[fluentui/react-popover] --> FC
    FC --> FIS[fluentui/react-image-storybook]
    FC --> FBS[fluentui/react-button-storybook]
    FC --> FPS[fluentui/react-popover-storybook]
```

Reply:
> This option should also result in a smaller package graph?

partially yes. 

if we switched to import only from packages instead of suite in stories, `react-components` suite build will be always triggered as it is in the `affected` graph (which will trigger builds for all RC deps). So it would introduce smaller "build" bottleneck per say.


to sum it up: 

while the suite architecture has consumer facing benefits (tree-shaking, no-dupes in node_modules), it inflicts DX issues ( makes building on any change basically almost everything )  -> that's why it's extremely important for us to have fast tooling in place. With pipeline caching in place the bottleneck situation should be also improved as only changed leaf packages will need to physically build ( other build will be fetched from cache )